apis/core/v1alpha1/clusterobjectsetphase_types.go

```diff
@@ -35,20 +40,33 @@ type ClusterObjectSetPhaseSpec struct {
 	// Immutable fields below
 
 	// Revision of the parent ObjectSet to use during object adoption.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="revision is immutable"
 	Revision int64 `json:"revision"`
 
 	// Previous revisions of the ClusterObjectSet to adopt objects from.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="previous is immutable"
+	// +kubebuilder:MaxItems=32
 	Previous []PreviousRevisionReference `json:"previous,omitempty"`
 
 	// Availability Probes check objects that are part of the package.
 	// All probes need to succeed for a package to be considered Available.
 	// Failing probes will prevent the reconciliation of objects in later phases.
+	// +kubebuilder:validation:Optional
```

Comment:
I added limits to the size of some of the fields for CEL budgeting. However, I have no idea what values actually make sense. Please suggest a more reasonable size limit.